client/navigation/utils.ts

```diff
@@ -43,50 +82,69 @@ export const getMatchScore = ( location, itemUrl, itemExpression = null ) => {
 	return ( decodeURIComponent( href ).match( regexp ) || [] ).length;
 };
 
-/**
- * Get a default expression to match the path and provided params.
- *
- * @param {string} url URL to match.
- * @return {string} Regex expression.
- */
-export const getDefaultMatchExpression = ( url ) => {
-	const escapedUrl = url.replace( /[-\/\\^$*+?.()|[\]{}]/gi, '\\$&' );
-	const [ path, args, hash ] = escapedUrl.split( /\\\?|#/ );
-	const hashExpression = hash ? `(.*#${ hash }$)` : '';
-	const argsExpression = args
-		? args.split( '&' ).reduce( ( acc, param ) => {
-				return `${ acc }(?=.*[?|&]${ param }(&|$|#))`;
-		  }, '' )
-		: '';
-	return '^' + path + argsExpression + hashExpression;
-};
+interface wcNavigation {
+	menuItems: Item[];
+	rootBackLabel: string;
+	rootBackUrl: string;
+	historyPatched: boolean;
+}
+
+declare global {
+	interface Window {
+		wcNavigation: wcNavigation;
+	}
+}
+
+interface wcNavigation {
+	menuItems: Array< Item >;
+	rootBackLabel: string;
+	rootBackUrl: string;
+	historyPatched: boolean;
+}
+
+declare global {
+	interface Window {
+		wcNavigation: wcNavigation;
+	}
+}
 
 /**
  * Adds a listener that runs on history change.
  *
  * @param {Function} listener Listener to add on history change.
  * @return {Function} Function to remove listeners.
  */
-export const addHistoryListener = ( listener ) => {
+export const addHistoryListener = ( listener: () => void ): ( () => void ) => {
 	// Monkey patch pushState to allow trigger the pushstate event listener.
 	if ( ! window.wcNavigation.historyPatched ) {
 		( ( history ) => {
 			/* global CustomEvent */
 			const pushState = history.pushState;
 			const replaceState = history.replaceState;
-			history.pushState = function ( state ) {
-				const pushStateEvent = new CustomEvent( 'pushstate', {
-					state,
-				} );
+			history.pushState = function (
+				state: {
+					[ key: string ]: string;
```

Comment:
Not a blocker but there is a built in type that can be useful for this: `Record<string, string>`